OP-952: Switch toogle and user roles picker synchronized

olewandowski1 opened the pull request:
Please do not merge it before [OTC-744 PR](https://github.com/openimis/openimis-fe-admin_js/pull/34) is not merged. 

The removal of region/district pickers in EO form was implemented in [OTC-744](https://openimis.atlassian.net/browse/OTC-744).

In scope of this ticket, I changed the way EO role picker and Switch element is connected.

sniedzielski commented on src/components/EnrolmentOfficerFormPanel.js:
```diff
@@ -42,19 +41,36 @@ const EnrolmentOfficerFormPanel = (props) => {
+  const isValid = !isLoading;
+
+  useEffect(() => {
+    const roles = edited.roles ? edited.roles : [];
```
Can you try with optional chaining? 
I think we can handle this by
`const roles = edited?.roles ?? [];`
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining

olewandowski1 replied:
Changed.

sniedzielski commented on src/components/EnrolmentOfficerFormPanel.js:
```diff
@@ -22,16 +22,15 @@ const styles = (theme) => ({
-      `
+  const [isEnabled, setIsEnabled] = useState(false);
+  const hasOfficerUserType = edited.userTypes?.includes(ENROLMENT_OFFICER_USER_TYPE);
+  const hasOfficerRole = edited.roles ? edited.roles.filter((x) => x.isSystem === 1).length !== 0 : false;
```
I wonder if we can also shortcut this line of code by using 'optional chaining'
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining

olewandowski1 replied:
Change of this condition have other consequences, so I decided to keep it like that.

sniedzielski reviewed (commented):
LGTM, I left two comments with suggestion to use optional chaining.

olewandowski1 commented:
In scope of last [commit](https://github.com/openimis/openimis-fe-admin_js/pull/35/commits/784a27494e2c1b94b56b02987a15b38abe4bd34f) I implemented the synchronization of adding CA role manually and CA Switch button. It was a case of [OP-1055](https://openimis.atlassian.net/browse/OP-1055).

I pushed it as a commit, because it strongly relates to the [OP-952](https://openimis.atlassian.net/browse/OP-952), where I did the identical logic in EO form. The other reason is fact that it is still not merged into [OTC-744].(https://openimis.atlassian.net/browse/OTC-744).

As I mentioned before, changes are virtually the same like, but could you please @sniedzielski look into it and review it?

olewandowski1 commented:
In scope of last [commit](https://github.com/openimis/openimis-fe-admin_js/pull/35/commits/784a27494e2c1b94b56b02987a15b38abe4bd34f) I implemented the synchronization of adding CA role manually and CA Switch button. It was a case of [OP-1055](https://openimis.atlassian.net/browse/OP-1055).

I pushed it as a commit, because it strongly relates to the [OP-952](https://openimis.atlassian.net/browse/OP-952), where I did the identical logic in EO form. The other reason is fact that it is still not merged into [OTC-744](https://openimis.atlassian.net/browse/OTC-744).

As I mentioned before, changes are virtually the same, but could you please @sniedzielski look into it and double check?

sniedzielski commented on src/components/ClaimAdministratorFormPanel.js:
```diff
@@ -14,35 +14,55 @@ const styles = (theme) => ({
+
+    if (isValid && isEnabled && !hasClaimRole) {
+      roles.push(claimRole);
+      edited.roles = roles;
```
@olewandowski1 seems that this useEffect looks quite the same as in '[EnrolmentOfficerFormPanel.js](https://github.com/openimis/openimis-fe-admin_js/pull/35/files#diff-dc8704c04aa3f4a8c4f2f8a9fe8d9cc46076651bf0b6e67d48f726447f89e9c3)'.
Can we create separate function to use it within those useEffects? I think this could be easy when it comes to the parametrize.

olewandowski1 replied:
Done.

sniedzielski commented on src/components/ClaimAdministratorFormPanel.js:
```diff
@@ -14,35 +14,55 @@ const styles = (theme) => ({
+    }
+  }, [isEnabled]);
+
+  useEffect(() => {
```
@olewandowski1 Same as in previous comment - looks like this useEffect is quite similar  as in '[EnrolmentOfficerFormPanel.js](https://github.com/openimis/openimis-fe-admin_js/pull/35/files#diff-dc8704c04aa3f4a8c4f2f8a9fe8d9cc46076651bf0b6e67d48f726447f89e9c3)'.
Can we create separate function to use it within those useEffects? I think this could be easy when it comes to the parametrize.

olewandowski1 replied:
Done.

olewandowski1 commented:
@sniedzielski I did what you've mentioned above. I created 2 functions in utils.js and I used it in both components. Could you please verify the solution?

sniedzielski reviewed (approved):
LGTM @olewandowski1

olewandowski1 commented:
@sniedzielski Thx for approval, I implemented changes mentioned in [OTC-744 PR](https://github.com/openimis/openimis-fe-admin_js/pull/34). Moreover, I changed variable names in `utils.js` to be more universal and clear.

delcroip commented:
Please @olewandowski1 fix the conflicts

dborowiecki reviewed (approved):
Looks good, thanks for resolving conflicts.
